Python MPL Visualization modules

codingS3b opened the pull request:
Supersedes #2479

ax3l commented on lib/python/picongpu/visualization/energy_histogram_visualizer.py:
```diff
@@ -0,0 +1,120 @@
+from picongpu.plugins.energy_histogram import EnergyHistogram
```
uh, forgot to mention: please add the copyright line to each file (see e.g. the data readers)

codingS3b commented:
Could you please have a look at the axis labels and plot titles for the single visualizers?

ax3l commented on lib/python/picongpu/plugins/plot_mpl/energy_histogram_visualizer.py:
```diff
@@ -0,0 +1,142 @@
+        self.plt_obj = None
+
+    def visualize(self, iteration, ax, species='e', species_filter="all",
+                  include_overflow=False, **kwargs):
```
I would remove `include_overflow` as an option since it will not be well visualized if it is included

codingS3b replied:
Would you completely remove it or be able to pass it down to the reader as kwargs? So would you want
```
counts, energy_bins = self.energy_histogram.get(
            species=species,
            species_filter=species_filter,
            iteration=iteration,
            **kwargs)
```
instead of
```
counts, energy_bins = self.energy_histogram.get(
            species=species,
            species_filter=species_filter,
            iteration=iteration,
            include_overflow=include_overflow)
```

ax3l replied:
For this specific visualization, which is not great to investigate overflows, I would say we just completely remove it.
If it's easy to pass in kwargs and that's the right place I won't stop you from allowing it there :)

ax3l commented on lib/python/picongpu/plugins/plot_mpl/energy_histogram_visualizer.py:
```diff
@@ -0,0 +1,142 @@
+License: GPLv3+
+"""
+
+
```
one newline less would be ok here :)

ax3l commented on lib/python/picongpu/plugins/plot_mpl/phase_space_visualizer.py:
```diff
@@ -0,0 +1,158 @@
+
+        # for unit-conversion from SI (taken from picongpu readthedocs)
+        self.mu = 1.e6
+        self.e_mc_r = 1. / (9.1e-31 * 2.9979e8)
```
if we want beta_gamma here, I have to check if we already specify the species' mass.
also, beta gamma per nucleon for ions would be useful

ax3l replied:
or we just show in units of MeV / c for now...

ax3l commented on lib/python/picongpu/plugins/plot_mpl/phase_space_visualizer.py:
```diff
@@ -0,0 +1,158 @@
+                "python", sys.argv[0],
+                "-p <path to run_directory> -i <iteration>"
+                " -s <particle species> -f <species_filter>"
+                " -m <momentum term>")
```
`momentum term` is actually both the `spatial component` and the `momentum component`.

`-m` is ok with me, but add `-m <phase space selection>` as in the data api

ax3l commented on lib/python/picongpu/plugins/plot_mpl/energy_histogram_visualizer.py:
```diff
@@ -0,0 +1,142 @@
+        else:
+            self.plt_obj.set_data(energy_bins, counts)
+
+        ax.set_xlabel('Energy')
```
'Energy  [keV]'

ax3l commented on lib/python/picongpu/plugins/plot_mpl/energy_histogram_visualizer.py:
```diff
@@ -0,0 +1,142 @@
+        ax.set_xlabel('Energy')
+        ax.set_ylabel('Count')
+        ax.set_title('Energy Histogram for species ' +
+                     species + ', iteration ' + str(iteration))
```
maybe add info on `species_filter` after species

ax3l commented on lib/python/picongpu/plugins/plot_mpl/energy_histogram_visualizer.py:
```diff
@@ -0,0 +1,142 @@
+                "python", sys.argv[0],
+                "-p <path to run_directory> -i <iteration>"
+                " -s <particle species> -f <species_filter>"
+                " -o <include_overflow>")
```
remove overflow option for viz

ax3l commented on lib/python/picongpu/plugins/plot_mpl/phase_space_visualizer.py:
```diff
@@ -0,0 +1,158 @@
+                '$')
+
+        ax.set_xlabel(r'${0}$ [${1}$]'.format(meta.r, "\mathrm{\mu m}"))
+        ax.set_ylabel(r'$p_{0}$ [$\beta\gamma$]'.format(meta.p))
```
unit should be `[$\beta\gamma m_\mathrm{species} / m_\mathrm{e}$]` right now...

ax3l commented on lib/python/picongpu/plugins/plot_mpl/energy_histogram_visualizer.py:
```diff
@@ -0,0 +1,131 @@
+        self.energy_histogram = EnergyHistogram(run_directory)
+        self.plt_obj = None
+
+    def visualize(self, iteration, ax, species='e', species_filter="all",
```
would be cool if the argument order would be as in the data api: species, species_filter, iteration, ax.

of course we can also invert additional options and iterations in both, but that should happen at once to be consistent

codingS3b replied:
Hm, this might be a problem for us, since as of right now we are using the visualizers in the following way (as a callback function of a jupyter-slider widget):

```{python}
iteration = change["new"]
idx = 0
for plugin in self.plugin_list:
    # each plugin gets an axes object for plotting
    # the sim data resides in a subdir of the experiment called \
    # 'run'
    plugin.visualize(iteration,
        self.ax_lst[idx],
        **self.layout_config)
        idx += 1
```
and the `layout_config` contains entries the user specifies (e.g. the species, filter and momentum type). Only this order allowed us to use this uniform way of calling the visualizers since they expect different optional parameters besides the iteration and the ax.
And the PNG Visualizer has (so far, at least) no options for filter or species (since I don't know whether you have PNG files for something other than electrons). 
All these points call for a common visualizer base class, what do you think? Can we create a common interface for all kinds of visualzers?

ax3l replied:
Sounds reasonable, yes.

codingS3b replied:
By the way, the PNG visualizer does all its reading on its own and has no reader class like the others...maybe that should be split as well...

codingS3b replied:
Do you have png visualization also for other species than electrons? Does species_filter make sense for png visualization as well?
Only then we could use a base class for visualizers with the signature
```python
class BaseVisualizer
    def visualize(self, species, species_filter, iteration, **kwargs)
```

ax3l replied:
yes, [the PNG plugin](https://picongpu.readthedocs.io/en/latest/usage/plugins/png.html#usage-plugins-png) can be used for any species and potentially in the future also `species_filters` (currently always `all`).

ax3l replied:
Please be aware that not all visualizers are species visualizers.

Our sources can be species, [(electro-magnetic) fields](https://picongpu.readthedocs.io/en/latest/usage/plugins/sliceFieldPrinter.html#usage-plugins-slicefieldprinter) or something totally different, e.g. a [scalar of total energy](https://picongpu.readthedocs.io/en/latest/usage/plugins/energyFields.html#usage-plugins-energyfields).

codingS3b replied:
I see... so what would you think could be an interface that is 
a) general enough and 
b) also reflects the order of arguments of the data readers?

ax3l replied:
probably the only common argument is `ax`. Although one could also query things globally for a simulation, in our case `iteration` could also be useful (maybe supporting ranges as well).

codingS3b replied:
When you have a png file, like `e_png_yz_0.5_006600.png`, what does the 0.5 stand for and what would be good parameter names for data readers for the coord label (yz) and that parameter behind the 0.5?
I'm asking because I'm trying to build a png reader as well

ax3l replied:
`xy` is the plane that is cut out and `0.5` is the relative offset ("slicePoint") in the third, here `z` "axis" (in the box limit: [0.:1.]):
https://picongpu.readthedocs.io/en/latest/usage/plugins/png.html#usage-plugins-png

ax3l commented on lib/python/picongpu/plugins/plot_mpl/energy_histogram_visualizer.py:
```diff
@@ -0,0 +1,131 @@
+        ----------
+        iteration: int
+            the iteration number for which data will be plotted.
+        ax: matplotlib axes object
```
a default for `ax` could be `plt.gca()`

codingS3b replied:
This default leads to the creation of another empty plot even when an ax argument is given, so I don't think we should do that...

ax3l replied:
oh, ok.
I was just thinking you do something like:
```python
def func(ax=None, ...)
    if ax is None:
        ax = plt.gca()
```

codingS3b replied:
yeah, that's what came to my mind as well...I'll add that

ax3l commented on lib/python/picongpu/plugins/plot_mpl/phase_space_visualizer.py:
```diff
@@ -0,0 +1,160 @@
+        self.mu = 1.e6
+        self.e_mc_r = 1. / (9.1e-31 * 2.9979e8)
+
+    def visualize(self, iteration, ax, species='e', species_filter='all',
```
order: see comment above

ax3l commented on lib/python/picongpu/plugins/plot_mpl/phase_space_visualizer.py:
```diff
@@ -0,0 +1,160 @@
+        ----------
+        iteration: int
+            the iteration number for which data will be plotted.
+        ax: matplotlib axes object
```
default: see comment above

ax3l commented on lib/python/picongpu/plugins/plot_mpl/png_visualizer.py:
```diff
@@ -0,0 +1,128 @@
+        ----------
+        iteration: int
+            the iteration number for which data will be plotted.
+        ax: matplotlib axes object
```
default: see comment above

ax3l commented:
@codingS3b when you next time push to this PR, please rebase against `dev` beforehand. We updated the travis scripts.

codingS3b commented:
I now updated the visualization module and added a first idea of a `base_visualizer` which all others inherit from and also moved parts of the `PNGVisualizer` into a `PNG` reader class.

ax3l commented:
can you please rebase against the latest `dev` and force-push your branch?

ax3l commented on lib/python/picongpu/plugins/plot_mpl/base_visualizer.py:
```diff
@@ -0,0 +1,44 @@
+
+        if run_directory is None:
+            raise ValueError('The run_directory parameter can not be None!')
+
```
maybe add a filesystem check if the dir exists?

codingS3b replied:
Isn't it sufficient that the reader performs this check? From what I saw they all check for existence in their constructors.

ax3l commented on lib/python/picongpu/plugins/plot_mpl/base_visualizer.py:
```diff
@@ -0,0 +1,44 @@
+
+class Visualizer(object):
+    def __init__(self, run_directory):
+
```
`run_directory` parameter: please add [the usual docstring](https://github.com/ComputationalRadiationPhysics/picongpu/blob/2c8a2bdd74ccf56c3db78c62d4d33c58110b3533/lib/python/picongpu/plugins/phase_space.py#L62-L68)

ax3l commented on lib/python/picongpu/plugins/plot_mpl/base_visualizer.py:
```diff
@@ -0,0 +1,44 @@
+    def _create_data_reader(self, run_directory):
+        raise NotImplementedError
+
+    def _create_plt_obj(self, ax):
```
what is the matplotlib baseclass this "plot_obj" refers to?
Do you mean an [`matplotlib.artist.Artist`](https://matplotlib.org/api/artist_api.html#matplotlib.artist.Artist)?

* https://matplotlib.org/users/artists.html

(I checked `plt.plot` which returns a list of Line2D which is a child class of `matplotlib.artist.Artist`)

codingS3b replied:
honestly I have not looked at whether all those matplotlib plotting objects have a common baseclass...What I meant with `self.plt_obj` was any object coming from matplotlib for which you could update data...

ax3l replied:
`plt.imshow` also returns an `matplotlib.artist.Artist`, so let use be explicit here since we are mpl only.

codingS3b replied:
But its always some function of the axes that we call, e.g. `ax.imshow()` or `ax.semilogy()` and they all seem to return different kinds of objects

ax3l commented on lib/python/picongpu/plugins/plot_mpl/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+matplotlib
+numpy
```
pls include this file in the `requirements.txt` a level above

codingS3b replied:
Sorry, I'm not sure how to write that correctly, do you mean `plugins/requirements.txt` would then look like this?
```
numpy
pandas>=0.21.0
h5py
matplotlib
-r .plot_mpl/requirements.txt
```

ax3l replied:
Close, just without the dot:

```bash
$ cat requirements.txt
-r lib/python/picongpu/requirements.txt

$ cat lib/python/picongpu/requirements.txt
-r input/requirements.txt
-r plugins/requirements.txt
-r utils/requirements.txt

$ cat lib/python/picongpu/plugins/requirements.txt
numpy
pandas>=0.21.0
h5py
matplotlib 
-r plot_mpl/requirements.txt
```

codingS3b replied:
Thanks, added this line now.

ax3l commented on lib/python/picongpu/plugins/plot_mpl/base_visualizer.py:
```diff
@@ -0,0 +1,96 @@
+        A matplotlib.axes.Axes instance
+        """
+
+        return ax or plt.gca()
```
I don't understand this. Won't this always just return the parameter `ax`?
Shouldn't it read
```python
if ax is None:
    return plt.gca()
else:
    return ax
````
?

codingS3b replied:
From my understanding the code that I wrote is just a shorter version of your suggestion (or at least I wanted it to be). When `ax=None` this will return the second variable, in this case `plt.gca()`. I tested it on ipython and it worked as expected.

ax3l replied:
Wow. Magic.

ax3l commented on lib/python/picongpu/plugins/png.py:
```diff
@@ -0,0 +1,207 @@
+"""
+
+import os
+import matplotlib.image as mpimg
```
I wonder if there is an alternative way to read in the PNG.
Would be in principle cool if the "data readers" would not need `matplotlib` but only the plotters...

anyway, ok for now with me.

codingS3b replied:
Using scipy would work without a lot of changes. Converting to a bytearray seems to loose the shape of the image matrix: when feeding that bytearray into matplotlib it complains about invalid dimensions for image data.
So if scipy as a dependency would be okay for you then I will change that

ax3l replied:
I checked the `scipy` library (not the [SciPy ecosystem/software stack](https://www.scipy.org/about.html)) dep list and it looks quite small. yes, let's do this.

codingS3b commented on lib/python/picongpu/plugins/energy_histogram.py:
```diff
@@ -100,8 +100,8 @@ def get_iterations(self, species, species_filter="all"):
 
-    def get(self, species, species_filter="all", iteration=None,
-            include_overflow=False):
+    def get(self, species=None, species_filter="all", iteration=None,
```
No its not, we can live without it, I just checked it.

ax3l replied:
ok cool, then please leave this two files unchanged :)

codingS3b commented on lib/python/picongpu/plugins/phase_space.py:
```diff
@@ -188,7 +188,8 @@ def get_iterations(self, species, species_filter='all', ps=None):
         return iterations
 
-    def get(self, species, species_filter='all', ps=None, iteration=None):
+    def get(self, species=None, species_filter='all', iteration=None, ps=None,
```
See above

codingS3b replied:
Is it still okay to change the order of the variables here? I recall you wanted to have species, filter and iteration before all other class dependent options so that is why I made the change.

ax3l replied:
Good question here. Well, ok.

If you change it, please adjust the `.rst` docs in `docs/source/usage/plugins/phaseSpace.rst` as well.

ax3l commented:
@codingS3b Sorry forgot to ping you, ready for rebase :)

codingS3b commented:
Final rebase was done!

ax3l commented:
thank you very much, great work! :rocket: :sparkles:
